emrichen/tags/loop.py

```diff
@@ -54,22 +54,35 @@ def enrich(self, context):
         index_as = str(self.data.get('index_as') or '')
         compact = bool(self.data.get('compact'))
         index_start = self.data.get('index_start')
+        previous_as = str(self.data.get('previous_as') or '')
 
         template = self.data.get('template')
         if template is None:
             raise ValueError('{self}: missing template'.format(self=self))
 
         output = []
         iterable, _ = self.get_iterable(context, index_start)
+        previous_value = None
         for index, value in iterable:
             subcontext = {as_: value}
             if index_as:
                 subcontext[index_as] = index
-            value = Context(context, subcontext).enrich(template)
-            if compact and not value:
+            if previous_as:
+                subcontext[previous_as] = previous_value
+            previous_value = value
+            subcontext = Context(context, subcontext)
+            result = subcontext.enrich(template)
+            if compact and not result:
                 continue
-            output.append(value)
+            self.process_item(subcontext, value, result)
+            output.append(result)
         return output
 
     def get_iterable(self, context, index_start):
         return get_iterable(self, self.data.get('over'), context, index_start)
+
+    def process_item(self, context, value, result):
+        '''
+        Used by Loop subclasses to do things every iteration.
+        '''
+        pass
```

Comment:
I feel this should be able to return an all-new `result`.

That is, `output.append(self.process_item(...))` above in `enrich()`, and `return result` as the default implementation here.

Reply:
Agree, a good improvement. However, if we do that, whose responsibility is to

1. `context.enrich(template)`? Still `Loop.enrich`?
2. implement `compact`?

OTOH, `output` is actually only used by `Loop`. `Index` and `Group` implement their own. Separate `BaseLoop`?